jupyterhubutils/authenticator/lsstjwtloginhandler.py

```diff
@@ -0,0 +1,133 @@
+'''LSST Login Handler to use JWT token present in request headers.
+'''
+import datetime
+from tornado import gen, web
+from jupyterhub.utils import url_path_join
+from jwtauthenticator.jwtauthenticator import JSONWebTokenLoginHandler
+
+
+class LSSTJWTLoginHandler(JSONWebTokenLoginHandler):
+
+    @gen.coroutine
+    def get(self):
+        '''Authenticate on get() via reading the token from HTTP headers.
+        '''
+        # This is taken from https://github.com/mogthesprog/jwtauthenticator
+        #  but with our additional claim information checked and stuffed
+        #  into auth_state, and allow/deny lists checked.
+        claims, token = self._check_auth_header()
+        username_claim_field = self.authenticator.username_claim_field
+        username = self.retrieve_username(claims, username_claim_field)
+        user = self.user_from_username(username)
+        # Here is where we deviate from the vanilla JWT authenticator.
+        # We simply store all the JWT claims in auth_state, although we also
+        #  choose our field names to make the spawner reusable from the
+        #  OAuthenticator implementation.
+        auth_state = yield self.refresh_user(user)
+        _ = yield user.save_auth_state(auth_state)
+        # Push the refreshed user through the managers
+        if not self._check_groups_jwt(claims):
+            # We're either in a forbidden group, or not in any allowed group
+            self.log.error("User did not validate from claims groups.")
+            raise web.HTTPError(403)
+        self.set_login_cookie(user)
+        _url = url_path_join(self.hub.server.base_url, 'home')
+        next_url = self.get_argument('next', default=False)
+        if next_url:
+            _url = next_url
+        self.redirect(_url)
+
+    @gen.coroutine
+    def refresh_user(self, user, handler=None):
+        '''Validate the token and force re-auth if the claims are not
+        (presumably no longer) valid.
+        '''
+        self.log.debug("Refreshing user data.")
+        try:
+            claims, token = self._check_auth_header()
+        except web.HTTPError:
+            # Force re-login
+            return False
+        username_claim_field = self.authenticator.username_claim_field
+        username = self.retrieve_username(claims, username_claim_field)
+        auth_state = {"id": username,
+                      "access_token": token,
+                      "claims": claims}
+        return auth_state
+
+    def _check_auth_header(self):
+        # Either returns (valid) claims and token,
+        #  or throws a web error of some type.
```

Comment:
maybe see if super class can do this?

Reply:
Nope, it's all jammed into the get() method and issues a redirect at the end.